### DISCUSSION TITLE
Creating a pivot table with average aggregation

### DISCUSSION THREAD

**Discussion by slhck** [saulpw/visidata#1535] [Q&A] [answered]
I'm struggling to do something which should be quite simple.

I have a dataset consisting of a worker name and its execution time:

```
ElasticsearchTransferSubWorker  27.055
DataFetchSubWorker      2.743
ElasticsearchTransferSubWorker  27.114
ElasticsearchTransferSubWorker  27.138
ElasticsearchTransferSubWorker  27.109
ElasticsearchTransferSubWorker  16.174
ElasticsearchTransferSubWorker  16.135
DataFetchSubWorker      2.55
ElasticsearchTransferSubWorker  27.156
ElasticsearchTransferSubWorker  15.62
ElasticsearchTransferSubWorker  16.709
ElasticsearchTransferSubWorker  15.799
ElasticsearchTransferSubWorker  16.89
DataFetchSubWorker      2.491
ElasticsearchTransferSubWorker  16.323
ElasticsearchTransferSubWorker  16.56
ElasticsearchTransferSubWorker  16.742
DataFetchSubWorker      3.79
ElasticsearchTransferSubWorker  28.297
ElasticsearchTransferSubWorker  28.172
```

I want to create a table that shows me the average execution time per worker:

```
ElasticsearchTransferSubWorker  XXX
DataFetchSubWorker XX
```

So I do:

* pipe into `vd --header 0`
* hit `!` on the first column
* hit `%` for the second column
* type `+avg`
* select the first column again
* type `Shift-W`

I get this:

![image](https://user-images.githubusercontent.com/582444/191687769-94a65523-6a59-4a26-bd1c-1f499992a4bd.png)

The rows are not collapsed as I would expect. Is there any way to accomplish this?

**Comment by saulpw** [comment, marked as answer]
Hi @slhck, you're very close.  Use `Shift+F` (Frequency Table) on the first column instead of `Shift+W` (Pivot), and it should make the sheet you want.

If you do want a horizontal pivot table, then don't press `!` to make it a key column; you just press `Shift+W` on the column and it should do what you want.

You also don't need `!` in the first case either, but it's harmless there.  It should probably be harmless in this case too. Thanks for bringing it to our attention!

**Reply by slhck** [reply]
Thanks, that did the trick!

I feel like the term *Frequency Table* made me ignore that part entirely, because I didn't want frequencies (i.e., counts); I knew I wanted a pivot table of some sort, because that's what I would have done in a regular spreadsheet. And all guides for pivot tables mention using `!`. Why is the `!` making it look so different though?

Not sure how this could be addressed in the documentation — arguably I haven't spent much time with the tool either —, but I couldn't really find a good solution through a web search and the docs didn't quite help.

**Reply by saulpw** [reply]
> Why is the ! making it look so different though?

It's trying to use the key column to determine the rows of the pivot table.  Without a key column, a Pivot (which will turn a long table into a wide one; it "pivots") will result in only one row.  If you use Pivot on a non-key-column, then each key value will have its own row, and each value in the pivoted column will have its own column (or multiple if there are mulitple aggregations set on it).  In your case, where the pivot column and the key column are the same, it should do what it did, but keep the key columns on the pivot sheet; then it would be more obvious that it's just a diagonal matrix.  The fact that it's omitting the key column on the pivot sheet is a bug of sorts (though maybe you could argue that it should just fail completely in this case).

Note that, while you think of the operation you wanted as a "pivot", it's actually not, because the ultimate table you want isn't "wide".  It's actually a "group by" in SQL terms; and the Frequency Table is the main way to do group bys (with default aggregators of count and percent of total).  Both Frequency Table and Pivot Table do a "group by", but as we generally want to keep tables long instead of wide in VisiData (and data in general), the Frequency Table is 10x more frequently (hah) useful than the Pivot Table.  This is just how it is.  If you've got ideas for how we can clear this up for new users like yourself (bearing in mind that discoverability of docs like this is a harder problem than writing the docs in the first place).

**Reply by slhck** [reply]
That makes sense, thanks! I was indeed thinking about an SQL “group by”, also coming from an R (tidyverse) background where I'd simply do `group_by(x) %>% summarize(mean(y))`.

I *guess* what would be useful is a basic juxtaposition of `visidata`'s approaches with what you might be used to from other tools (like Excel, SQL, `dplyr`,  `pandas`), although this would be a) a lot of work and b) doesn't make up for people not spending the required amount of time to work with a tool. ([Example of such a comparison here](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_r.html).)

With all that said, I think we've left enough keywords here for people to discover an answer should they be in a similar situation like me!
